Mathlib/Data/Rat/MetaDefs.lean

```diff
@@ -0,0 +1,161 @@
+/-
+Copyright (c) 2019 Robert Y. Lewis . All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Robert Y. Lewis
+
+! This file was ported from Lean 3 source module data.rat.meta_defs
+! leanprover-community/mathlib commit c18a48e9f71115845326e03443913f0c3694c153
+! Please do not edit these lines, except to modify the commit id
+! if you have ported upstream changes.
+-/
+import Mathlib.Data.Rat.Defs
+import Mathlib.Tactic.Core
+import Qq
+
+/-!
+# Meta operations on ℚ
+
+This file defines functions for dealing with rational numbers as expressions.
+
+They are not defined earlier in the hierarchy, in the `tactic` or `meta` folders, since we do not
+want to import `data.rat.basic` there.
+
+## Main definitions
+
+* `rat.mk_numeral` embeds a rational `q` as a numeral expression into a type supporting the needed
+  operations. It does not need a tactic state.
+* `rat.reflect` specializes `rat.mk_numeral` to `ℚ`.
+* `expr.of_rat` behaves like `rat.mk_numeral`, but uses the tactic state to infer the needed
+  structure on the target type.
+
+* `expr.to_rat` evaluates a normal numeral expression as a rat.
+* `expr.eval_rat` evaluates a numeral expression with arithmetic operations as a rat.
+
+-/
+
+open Qq
+
+/-- `Int.toExprQ α _ z _` embeds `q` as a numeral expression inside a type with `OfNat` and `-`.
+-/
+def Int.toExprQ {u : Lean.Level} (α : Q(Type u)) (_ : Q(Neg $α)) (z : ℤ)
+  (_ : let zna := z.natAbs; by exact Q(OfNat $α $zna)) : Q($α) :=
```

Comment:
@gebner, is there syntax along the lines of `Q(OfNat $α $(z.natAbs))`? It doesn't seem to work here.

Reply:
No, the antiquotations are assumed to be of type `Q(_)`.  Longer-term, it would probably be best to explicitly write `toExprQ z.natAbs` (bikeshed) here.

Reply:
Oh, I didn't know about `toExprQ` (I assume it already exists?)

Reply:
Not yet.  See #5952. :smile: